### PR TITLE
New service to re-sync documents with publishing-api

### PR DESCRIPTION
**Pull request by leenagupte**
Trello: https://trello.com/c/cT9yYO4X
Replaces PR #1487 

## What's changed and why?
Adds a service to re-sync a document with publishing-api. This will mainly be used in two scenarios:

1. Documents imported from Whitehall will need to be re-presented to publishing-api and have their base_path reservation moved from Whitehall to Content Publisher.

2. All documents will need to be represented to publishing-api to associate them with the correct government, so that if there is a new government, they can successfully transition into history mode.

co-authored with @ChrisBAshton

### CODE REVIEW

File: spec/services/resync_service_spec.rb [hunk 1]
```diff
@@ -50,5 +50,37 @@
...
+
+        withdraw_params = {
+          type: "withdrawal",
+          explanation: explanation,
+          locale: edition.locale,
+        }
+
+        expect(GovspeakDocument)
+          .to receive(:new)
+          .with(edition.status.details, edition)
```
Comment (kevindew) [inline spec/services/resync_service_spec.rb]: You can probably change this to:

```
        expect(GovspeakDocument)
          .to receive(:new)
          .with(edition.status.details, edition)
          .and_return(instance_double(GovspeakDocument, payload_html: explanation))
```
and remove the need for the allow on line 65

File: spec/factories/edition_factory.rb [hunk 2]
```diff
@@ -119,6 +119,25 @@
...
 
+    trait :removed do
+      live { true }
+
+      transient do
+        removal { nil }
+      end
+
+      after(:build) do |edition, evaluator|
+        edition.revision = evaluator.association(:revision)
```
Comment (kevindew) [inline spec/factories/edition_factory.rb]: I don't think we should or need to do this, it seems to be a mistake we have with withdrawn factory too.

File: spec/services/resync_service_spec.rb [hunk 3]
```diff
@@ -82,5 +82,58 @@
...
       end
     end
+
+    context "when the live edition has been removed" do
+      let(:explanation) { "explanation" }
+
+      before do
+        stub_any_publishing_api_unpublish
+        allow(PreviewAssetService).to receive(:call)
+        allow(PublishAssetService).to receive(:call)
```
Comment (kevindew) [inline spec/services/resync_service_spec.rb]: Do these stubs need to be here? not sure if they have an impact.

File: app/services/edit_edition_service.rb [hunk 4]
```diff
@@ -28,9 +28,6 @@ def determine_political
   end
 
   def associate_with_government
-    edition.government_id = if edition.public_first_published_at
-                              government = Government.for_date(edition.public_first_published_at)
-                              government&.content_id
-                            end
+    edition.government_id = Government.for_date(edition.public_first_published_at)&.content_id
```
Comment (kevindew) [inline app/services/edit_edition_service.rb]: Nice 👍

File: lib/tasks/resync.rake [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    document = Document.find_by_param(args.content_id_and_locale)
+
+    raise "No document exists for #{content_id_and_locale}" unless document
+
+    ResyncService.call(document)
+  end
+
+  desc "Resync all documents with the publishing-api e.g. resync:all"
+  task all: :environment do
+    Document.all.each do |document|
```
Comment (kevindew) [inline lib/tasks/resync.rake]: We should swap this to `Document.find_each` so it loads data in batches.

File: spec/lib/tasks/resync_spec.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        .with(document)
+
+      Rake::Task["resync:document"].invoke("#{document.content_id}:#{document.locale}")
+    end
+  end
+
+  describe "resync:all" do
+    it "resyncs all documentss" do
+      create(:document)
+      create(:document)
```
Comment (kevindew) [inline spec/lib/tasks/resync_spec.rb]: minor but you can do `create_list(:document, 2)`

File: spec/services/resync_service_spec.rb (hunk 3 above)
Comment (leenagupte) [reply spec/services/resync_service_spec.rb]: I was getting an because I'd set up the factory incorrectly. Will remove them.

**Comment by leenagupte** [comment]
@kevindew Thanks for the review. I've addressed your comments and fixed up.

**Comment by leenagupte** [comment]
> Shall we add an AC to the trello card about the gds-api-adapters change so we don't forget to do that?

New ticket added here: https://trello.com/c/NC10wudi/1276-add-a-test-helper-to-stub-the-putpath-method-in-gds-api-adapters

File: app/services/resync_service.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    end
+  end
+
+private
+
+  def sync_live_edition
+    live_edition.lock!
+    PreviewService.call(live_edition, republish: true)
+    PublishAssetService.call(live_edition, nil)
+    publish
```
[the remark this answers is not in the document]
Comment (leenagupte) [reply app/services/resync_service.rb]: I don't think the edition is being updated without a bang (!), so I don't think a `save!` is needed. Why do you think it's needed?

[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply app/services/resync_service.rb]: Ahh, you're right. We're calling `update!`, which calls `save!` under the hood. I was worried we weren't releasing the lock.

File: app/services/resync_service.rb [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+  def redirect_or_remove
+    removal = live_edition.status.details
+    GdsApi.publishing_api_v2.unpublish(
+      live_edition.content_id,
+      type: removal.redirect? ? "redirect" : "gone",
+      explanation: removal.explanatory_note,
+      alternative_path: removal.alternative_path,
+      locale: live_edition.locale,
+      allow_draft: true,
```
Comment (kevindew) [inline app/services/resync_service.rb]: We probably should have an unpublished_at here as well.

Comment (leenagupte) [reply app/services/resync_service.rb]: @kevindew I've updated and fixed up

**Pull request merged**